core-bundle/src/Routing/ResponseContext/ResponseContext.php

```diff
@@ -12,31 +12,72 @@
 
 namespace Contao\CoreBundle\Routing\ResponseContext;
 
-use Symfony\Component\HttpFoundation\Response;
-use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException;
 
-class ResponseContext implements ResponseContextInterface
+final class ResponseContext
 {
+    public const REQUEST_ATTRIBUTE_NAME = '_contao_response_context';
+
+    /**
+     * @var Container
+     */
+    private $container;
+
     /**
      * @var PartialResponseHeaderBag|null
      */
     private $headerBag;
 
-    public function getHeaderBag(): ResponseHeaderBag
+    public function __construct()
     {
-        if (null === $this->headerBag) {
-            $this->headerBag = new PartialResponseHeaderBag();
-        }
+        $this->container = new Container();
+    }
 
-        return $this->headerBag;
+    public function remove(string $serviceId): self
+    {
+        $this->container->set($serviceId, null);
+
+        return $this;
     }
 
-    public function finalize(Response $response): ResponseContextInterface
+    public function add(object $service): self
     {
-        foreach ($this->getHeaderBag()->all() as $name => $values) {
-            $response->headers->set($name, $values, false); // Do not replace but add
+        $this->container->set(\get_class($service), $service);
+
+        // Automatically add aliases for all interfaces (last one added automatically wins by overriding here)
+        foreach ((new \ReflectionClass($service))->getInterfaceNames() as $interfaceName) {
+            $this->container->set($interfaceName, $service);
```

Comment:
Should we add parent classes too, so that you can overwrite a service that doesn’t implement an interface?

```suggestion
            $this->container->set($interfaceName, $service);
        }
        
        $parent = new \ReflectionClass($service);
        while ($parent = $parent->getParentClass()) {
            $this->container->set($parent->getName(), $service);
```

Reply:
Hmmm. What about the parent of the parent? Isn't that an endless discussion?

Reply:
> What about the parent of the parent?

My suggested code already handles that ☺️

Reply:
This way we add all parents and all interfaces, so it functions exactly the same as type hints everywhere in PHP.

Reply:
Yeah I know, I just don't know if it's a good idea 🙈 Do you have a specific use case in mind and why we would want to support this? I mean, what's the point? If it's a concrete class, there's not two services that can extend and work at the same time. It would be better to use an interface and composition for that case (and even this case is highly unlikely to ever happen in the wild). We can still add this if there's a real use case.

Reply:
I just thought it would be more “complete” this way so that all possible `T`s  for `$service instanceof T` get set. As we now store all interfaces, all parent classes interfaces and all parent interfaces of these interfaces, I thought the parent classes are just the last one missing.

But I’m fine too with merging it as is.

Reply:
@contao/developers you decide. I think neither @ausi nor I do have strong preference here. We can add it now, we can also add it later.

Reply:
I would either add parents as well or no aliases at all.

Reply:
I am with @m-vo.